src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -0,0 +1,251 @@
+            {
+                throw new ArgumentNullException("request", SR.net_http_handler_norequest);
+            }
+            if (request.RequestUri.Scheme == UriSchemeHttps)
+            {
+                throw NotImplemented.ByDesignWithMessage("HTTPS stack is not yet implemented");
+            }
+            if (request.Content != null)
+            {
+                throw NotImplemented.ByDesignWithMessage("HTTP requests with a body are not yet supported");
+            }
+
+            CheckDisposed();
+
+            SetOperationStarted();
+
+            // Create RequestState object and save current values of handler settings.
+            RequestState state = new RequestState();
+            state.CancellationToken = cancellationToken;
+            state.RequestMessage = request;
+
+            try
+            {
+                Task.Factory.StartNew(
+                    StartRequest,
+                    state,
+                    CancellationToken.None,
+                    TaskCreationOptions.DenyChildAttach,
+                    TaskScheduler.Default);
+            }
+            catch (Exception ex)
+            {
+                throw new HttpRequestException("SendAsync", ex);
+            }
+
+            return state.Task;
+        }
+
+        #region Private methods
+
+        private void StartRequest(object obj)
+        {
+            RequestState state = (RequestState)obj;
+            HttpResponseMessage responseMessage = null;
+            Exception savedException = null;
+            SafeCurlHandle requestHandle = null;
+
+            if (state.CancellationToken.IsCancellationRequested)
+            {
+                state.TrySetCanceled();
+                return;
+            }
+
+            state.CancellationToken.Register(x => ((RequestState)x).TrySetCanceled(), state);
+
+            try
+            {
+                requestHandle = CreateRequestHandle(state);
+
+                state.CancellationToken.ThrowIfCancellationRequested();
+                int result = Interop.libcurl.curl_easy_perform(requestHandle);
+                if (CURLcode.CURLE_OK != result)
+                {
+                    throw new HttpRequestException(SR.net_http_client_execution_error, GetCurlException(result));
+                }
+
+                // TODO: Handle requests with body
+
+                state.CancellationToken.ThrowIfCancellationRequested();
+                responseMessage = CreateResponseMessage(requestHandle, state.RequestMessage);
+            }
+            catch (Exception ex)
+            {
+                savedException = ex;
+            }
+            SafeCurlHandle.DisposeAndClearHandle(ref requestHandle);
+
+            // Move the task to a terminal state.
+            if (responseMessage != null)
+            {
+                state.TrySetResult(responseMessage);
+            }
+            else
+            {
+                HandleAsyncException(state, savedException);
+            }
```

Comment:
It's not clear to me why the latter half of this method is structured this way, storing the exception and then processing it afterwards.  Why not change it to be:
```C#
    ...
    state.TrySetResult(responseMessage);
}
catch (Exception ex)
{
    HandleAsyncException(state, ex);
}
finally
{
    SafeCurlHandle.DisposeAndClearHandle(ref requestHandle);
}
```
?